core/arch/arm/kernel/thread_spmc.c

```diff
@@ -1843,27 +1848,51 @@ void thread_spmc_relinquish(uint64_t cookie)
 		EMSG("Failed to relinquish cookie %#"PRIx64, cookie);
 }
 
-static int set_pages(struct ffa_address_range *regions,
-		     unsigned int num_regions, unsigned int num_pages,
-		     struct mobj_ffa *mf)
+static int add_pages_at(struct ffa_address_range *regions,
+			unsigned int num_regions, unsigned int *idx,
+			struct mobj_ffa *mf)
 {
 	unsigned int n = 0;
-	unsigned int idx = 0;
 
 	for (n = 0; n < num_regions; n++) {
 		unsigned int page_count = READ_ONCE(regions[n].page_count);
 		uint64_t addr = READ_ONCE(regions[n].address);
 
-		if (mobj_ffa_add_pages_at(mf, &idx, addr, page_count))
+		if (mobj_ffa_add_pages_at(mf, idx, addr, page_count))
 			return FFA_INVALID_PARAMETERS;
 	}
 
-	if (idx != num_pages)
-		return FFA_INVALID_PARAMETERS;
-
 	return 0;
 }
 
+static struct ffa_mem_transaction *spmc_frag_retrieve(uint64_t cookie,
```

Comment:
struct ffa_mem_transaction or struct ffa_address_range ?

Reply:
Yes, I believe that's correct.

Reply:
ffa_mem_transaction doesn't like a good fit, or modify it to the （void *） type to be consistent with the function spmc_retrieve_req return type, what do you think?